contracts/DixelClubV2NFT.sol

```diff
@@ -143,50 +159,74 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
 
     // @dev Maximum length of list array can be limited by block gas limit of blockchain
     // @notice Duplicated address input means multiple allowance
-    function addWhitelist(address[] memory list) external onlyOwner {
-        require(_metaData.whitelistOnly, "COLLECTION_IS_PUBLIC");
+    function addWhitelist(address[] calldata list) external onlyOwner {
+        if(!_metaData.whitelistOnly) revert DixelClubV2__PublicCollection();
 
         uint256 length = list.length; // gas saving
-        for (uint256 i = 0; i < length; i++) {
-            _whitelist.push(list[i]);
+        address curr; // gas saving
+        for (uint256 i; i != length; ) {
+            curr = list[i];
+            PositionCount storage pc = _whitelistData[curr];
+
+            if (!_whitelistData[address(0)].hitted[curr]) {
+                _whitelistData[address(0)].hitted[curr] = true;
+                _whitelist.push(curr);
+                unchecked {
+                    (pc.position, pc.init) = (uint96(_whitelist.length) - 1, true);
+                }
+            }
+
+            unchecked {
+                ++pc.count;
+                ++i;
+            }
         }
+
+        delete _whitelistData[address(0)];
     }
 
     function _removeWhitelist(address wallet) private {
-        for (uint256 i = 0; i < _whitelist.length; i++) {
-            if (_whitelist[i] == wallet) {
-                _whitelist[i] = _whitelist[_whitelist.length - 1]; // put the last element into the delete index
-                _whitelist.pop(); // delete the last element to decrease array length;
-
-                break; // delete the first matching one and stop
-            }
+        PositionCount storage pc = _whitelistData[wallet];
+        if (--pc.count == 0 && pc.init == true) {
+            _whitelist[pc.position] = _whitelist[_whitelist.length - 1];
+            _whitelist.pop();
+            delete _whitelistData[wallet];
         }
     }
 
     // @dev Maximum length of list array can be limited by block gas limit of blockchain
-    function removeWhitelist(address[] memory list) external onlyOwner {
-        require(_metaData.whitelistOnly, "COLLECTION_IS_PUBLIC");
+    function removeWhitelist(address[] calldata list) external onlyOwner {
+        if(!_metaData.whitelistOnly) revert DixelClubV2__PublicCollection();
 
         uint256 length = list.length; // gas saving
-        for (uint256 i = 0; i < length; i++) {
+        for (uint256 i; i != length;) {
             _removeWhitelist(list[i]);
+            unchecked {
+                i++;
```

Comment:
Change to `++i` to take advantage of pre-increment operation.

Reply:
Oh i missed!
You are very Good reviewer!